src/codemodder/codemods/use_walrus_if.py

```diff
@@ -0,0 +1,130 @@
+        rules:
+          - patterns:
+            - pattern: |
+                $ASSIGN
+                if $COND:
+                  $BODY
+            - metavariable-pattern:
+                metavariable: $ASSIGN
+                patterns:
+                  - pattern: $VAR = $RHS
+                  - metavariable-pattern:
+                      metavariable: $COND
+                      patterns:
+                        - pattern: $VAR
+                  - metavariable-pattern:
+                      metavariable: $BODY
+                      pattern: $VAR
+            - focus-metavariable: $ASSIGN
+        """
+
+    _modify_next_if: List[Tuple[CodeRange, cst.Assign]]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._modify_next_if = []
+
+    def add_change(self, position: CodeRange):
+        self.file_context.codemod_changes.append(
+            Change(
+                lineNumber=position.start.line,
+                description=self.CHANGE_DESCRIPTION,
+            ).to_json()
+        )
+
+    def leave_If(self, original_node, updated_node):
+        if self._modify_next_if:
+            position, if_node = self._modify_next_if.pop()
+            is_name = m.matches(updated_node.test, m.Name())
+            named_expr = cst.NamedExpr(
+                target=if_node.targets[0].target,
+                value=if_node.value,
+                lpar=[] if is_name else [cst.LeftParen()],
+                rpar=[] if is_name else [cst.RightParen()],
+            )
+            self.add_change(position)
+            return (
+                updated_node.with_changes(test=named_expr)
+                if is_name
+                else updated_node.with_changes(
+                    test=updated_node.test.with_changes(left=named_expr)
+                )
+            )
+
+        return original_node
+
+    def _is_valid_modification(self, node):
+        """
+        Restricts the kind of modifications we can make to the AST.
+
+        This is necessary since the semgrep rule can't fully encode this restriction.
+        """
+        if parent := self.get_metadata(ParentNodeProvider, node):
+            if gparent := self.get_metadata(ParentNodeProvider, parent):
+                if (idx := gparent.children.index(parent)) >= 0:
+                    return m.matches(
+                        gparent.children[idx + 1],
+                        m.If(test=(m.Name() | m.Comparison(left=m.Name()))),
+                    )
+        return False
+
+    def leave_Assign(self, original_node, updated_node):
+        if self.node_is_selected(original_node):
+            if self._is_valid_modification(original_node):
+                position = self.node_position(original_node)
+                self._modify_next_if.append((position, updated_node))
+                return cst.RemoveFromParent()
+
+        return original_node
+
+    def leave_SimpleStatementLine(self, original_node, updated_node):
```

Comment:
This is interesting and I take your point, but I think that this would be very difficult to implement in terms of the visitor pattern since it would require quite a bit of state to be passed around. I can experiment with it a little longer but since this is already working with the behavior we want I might just have to revisit it later.

Reply:
I added a comment to the code to preserve this info.